File: src/main/java/com/github/jactor/web/JactorWeb.java
package com.github.jactor.web;

import static java.util.Arrays.stream;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.boot.CommandLineRunner;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.context.ApplicationContext;
import org.springframework.context.annotation.Bean;
import org.springframework.web.servlet.config.annotation.WebMvcConfigurer;

@SpringBootApplication
public class JactorWeb implements WebMvcConfigurer {

  private static final Logger LOGGER = LoggerFactory.getLogger(JactorWeb.class);

  @Bean
  public CommandLineRunner commandLineRunner(ApplicationContext applicationContext) {
    return args -> inspect(applicationContext, args);
  }

  private void inspect(ApplicationContext applicationContext, String[] args) {
    if (LOGGER.isDebugEnabled()) {
      boolean noArgs = args == null || args.length == 0;
      String arguments = noArgs ? "without arguments" : "with arguments: " + String.join(" ", args);

      LOGGER.debug("Starting jactor-web {}", arguments);

      SpringBeanNames springBeanNames = new SpringBeanNames();
      stream(applicationContext.getBeanDefinitionNames()).sorted().forEach(springBeanNames::add);

      if (LOGGER.isDebugEnabled()) {
        LOGGER.debug("Available beans:");
        springBeanNames.listBeanNames().stream().map(name -> "- " + name).forEach(LOGGER::debug);
      }

      if (LOGGER.isTraceEnabled()) {
        LOGGER.trace("Available spring beans:");
        springBeanNames.listSpringBeans().stream().map(name -> "- " + name).forEach(LOGGER::trace);
      }
    }
  }

  public static void main(String... args) {
    SpringApplication.run(JactorWeb.class, args);
  }
}

File: web/src/main/kotlin/com/github/jactor/web/JactorWeb.kt
package com.github.jactor.web

import org.springframework.boot.CommandLineRunner
import org.springframework.boot.SpringApplication
import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.context.ApplicationContext
import org.springframework.context.annotation.Bean
import org.springframework.web.servlet.config.annotation.WebMvcConfigurer
import com.github.jactor.shared.SpringBeanNames
import io.github.oshai.kotlinlogging.KotlinLogging

private val logger = KotlinLogging.logger {}

@SpringBootApplication
class JactorWeb : WebMvcConfigurer {
    @Bean
    fun commandLineRunner(applicationContext: ApplicationContext): CommandLineRunner {
        return CommandLineRunner { args: Array<String>? -> inspect(applicationContext, args ?: emptyArray()) }
    }

    private fun inspect(applicationContext: ApplicationContext, args: Array<String>) {
        logger.debug {
            logger.debug { "Starting jactor-web ${gatherArgs(args)}" }
            logger.debug { "Available beans (only simple names):" }

            SpringBeanNames().also { springBeanNames ->
                applicationContext.beanDefinitionNames.sorted().forEach(springBeanNames::add)
                springBeanNames.listBeanNames().forEach {
                    logger.debug { "- $it" }
                }
            }

            "Ready for service..."
        }
    }

    fun gatherArgs(args: Array<String>): String {
        val arguments = if (args.isEmpty()) {
            "without arguments!"
        } else {
            "with arguments: ${args.joinToString { " " }}!"
        }
        return arguments
    }
}

fun main(args: Array<String>) {
    SpringApplication.run(JactorWeb::class.java, *args)
}
